variables/variables.go

```diff
@@ -0,0 +1,292 @@
+package variables
+
+import (
+	"fmt"
+	"github.com/gruntwork-io/boilerplate/util"
+)
+
+// An interface for a variable defined in a boilerplate.yml config file
+type Variable interface {
```

Comment:
Unfortunately, the options for cross-platform CLI apps are very limited. Interpreted/dynamic languages like Ruby and Python are wonderful to work with, but a nightmare to distribute, as slightly different versions of Ruby, Python, and all their dependent libraries wreak havoc on portability. Compiled languages like Java and Scala include way too much overhead for simple CLI apps (e.g. you have to download 300 jars to run "Hello, World"). 

That leaves bash and Go. Bash is not a language built for anything over 100 lines, so it's not really an option for complicated apps. So that leaves Go.

The only other option is [Rust](https://www.rust-lang.org/en-US/). I haven't used it, but have heard very good things, and I *think* it can do very lightweight, cross-platform binaries. My only concern is that it's relatively new and it's community is much smaller than Go's, so finding good 3rd party libraries would be a pain. It's supposed to be a fairly elegant and modern language, and hopefully the future of systems programming (a C/C++ replacement), but I'm not sure that's enough to warrant a switch right now.